Allow string transforms to be applied to sections and conf keys

mreis1 opened the pull request:
Reads configuration files in a case insensitive way thanks to the transform method that I implemented. 

Some programming languages such as delphi reads 
[Section1] or [SeCtion1] as being the same section. They are case insensitive. 

In order to address this need I introduced transforms. 
{transform: 'upper' | 'lower' | 'none' }  (default: 'none')

[Section1] would be converted into [SECTION1]
And same to all configuration property's keys.

Commit:
convert keys to case insensitive by allowing us to apply transforms such as 'upper' or 'lower'

Commit:
0.2.7

Commit:
Remove console.log

ZachPerkitny commented on src/configparser.js:
```diff
@@ -231,22 +237,39 @@ ConfigParser.prototype.writeAsync = function(file) {
+	let transform = options.transform;
     let curSec = null;
+    let upperCaseKey = function(key){
+	    if (key === void 0 || key === null) return;
```
There seems to be an inconsistent use of spaces and tabs, could you use 4 space indentation please ?

ZachPerkitny commented on src/configparser.js:
```diff
@@ -231,22 +237,39 @@ ConfigParser.prototype.writeAsync = function(file) {
+        if (!transform){
+            return key;
+        }
+	    if (transform === 'upper') {
```
Maybe using a JS 'enum' by adding
```js
const Transform = Object.freeze({
    NONE: 0,
    UPPER: 1,
    LOWER: 2
});
```

```js
if (transform == Transform.UPPER) {
    return key.toUpperCase();
} else if (transform == Transform.LOWER) {
    return key.toLowerCase();
} else {
    return key;
}
```

ZachPerkitny commented on test/configparser.js:
```diff
@@ -5,17 +5,17 @@ describe('ConfigParser object', function(){
-            'more_complex_interpolation'
-        ]);
-    });
+	it('should return all sections in the config file', function () {
```
Use 4 spaces for indentation please.

ZachPerkitny commented on test/configparser.js:
```diff
@@ -80,4 +80,19 @@ describe('ConfigParser object', function(){
     });
+
+
+	describe('Transform configuration keys', function(){
```
Use 4 spaces for indentation please, I should really add a eslint file...

ZachPerkitny commented on src/configparser.js:
```diff
@@ -231,22 +237,39 @@ ConfigParser.prototype.writeAsync = function(file) {
+    options = options || {};
+	let transform = options.transform;
     let curSec = null;
+    let upperCaseKey = function(key){
```
This could be `const` and maybe rename it to `transformKey`

ZachPerkitny commented on src/configparser.js:
```diff
@@ -36,7 +36,9 @@ const writeFileAsync = util.promisify(fs.writeFile);
-function ConfigParser() {
+function ConfigParser(options) {
+	options = options || {};
+	this._transform = options.transform || 'none';
```
4 Space Indentation please, and we could just use default params, so `function ConfigParser(options = {}) {...}`

ZachPerkitny commented on src/configparser.js:
```diff
@@ -231,22 +237,39 @@ ConfigParser.prototype.writeAsync = function(file) {
-function parseLines(lines) {
+function parseLines(lines, options) {
+    options = options || {};
+	let transform = options.transform;
```
Use 4 space indentation.

ZachPerkitny commented on src/configparser.js:
```diff
@@ -231,22 +237,39 @@ ConfigParser.prototype.writeAsync = function(file) {
 }
 
-function parseLines(lines) {
+function parseLines(lines, options) {
```
we could just use default params, so `function parseLines(lines, options = {}) {...}`

ZachPerkitny reviewed (commented):
I've left feedback regarding indentation, naming and the use of JS 'enums' instead of strings.

mreis1 committed:
fix: lines starting with spaces leading to parsing error
fix: Parsing error being misleading since a variable "file" was undefined during the parseLine function and as consequence parse would fail with no details such as error line

mreis1 committed:
add onInvalidLine detection and make it possible to silently handle those errors
